### PR TITLE
Add ability to test demos of specific implementation.

### PR DESCRIPTION
**Pull request by ivikhrev**
For `--demos` argument `gapi, cpp, python` values were added, which enable testing demos of corresponding implementation. Added exception in case of empty list of demos to test.

### CODE REVIEW

**Review by Wovchena — CHANGES_REQUESTED** [review]

**Review by anzhella-pankratova — APPROVED** [review]
The introduction of performance scope affects this PR: if `--demos object_detection_demo/python` is passed, only test cases from performance_cases will be taken.  

Please, could you fix it in this PR:  in performance_cases when we modify demos from `BASE` dictionary, we should take deepcopy of value, like `deepcopy(BASE['object_detection_demo/python'])`

File: demos/tests/cases.py [hunk 1]
```diff
@@ -1392,6 +1394,7 @@ def single_option_cases(key, *args):
     )),
 ]
 
-DEMOS = NATIVE_DEMOS + PYTHON_DEMOS
 
+DEMOS = GAPI_DEMOS + CPP_DEMOS + PYTHON_DEMOS
+DEMOS_PER_IMPL = { 'gapi' : GAPI_DEMOS, 'cpp' : CPP_DEMOS, 'python' : PYTHON_DEMOS }
```
[the remark this answers is not in the document]
Comment (ivikhrev) [reply demos/tests/cases.py]: done

File: demos/tests/run_tests.py [hunk 2]
```diff
@@ -210,10 +212,26 @@ def main():
 
     if args.demos is not None:
         names_of_demos_to_test = set(args.demos.split(','))
+        if all(name in DEMOS_PER_IMPL.keys() for name in names_of_demos_to_test):
+            demos_of_specific_impls = []
+            for name in names_of_demos_to_test:
+                demos_of_specific_impls += DEMOS_PER_IMPL[name]
+            names_of_demos_to_test = [demo.subdirectory for demo in demos_of_specific_impls]
```
[the remark this answers is not in the document]
Comment (ivikhrev) [reply demos/tests/run_tests.py]: done

File: demos/tests/run_tests.py [hunk 3]
```diff
@@ -68,7 +69,8 @@ def parse_args():
     parser.add_argument('--downloader-cache-dir', type=Path, required=True, metavar='DIR',
         help='directory to use as the cache for the model downloader')
     parser.add_argument('--demos', metavar='DEMO[,DEMO...]',
-        help='list of demos to run tests for (by default, every demo is tested)')
+        help='list of demos to run tests for (by default, every demo is tested). '
+        'For testing demos of specific implementation pass one (or more) of the next values: gapi, cpp, python.')
```
[the remark this answers is not in the document]
Comment (ivikhrev) [reply demos/tests/run_tests.py]: done

File: demos/tests/cases.py [hunk 4]
```diff
@@ -29,9 +29,11 @@
 
 
 class Demo:
+    IMPLEMETATION_TYPES = set()
```
Comment (anzhella-pankratova) [inline demos/tests/cases.py:32]:
```suggestion
    IMPLEMENTATION_TYPES = set()
```

Comment (ivikhrev) [reply demos/tests/cases.py:32]: yeah, thanks

File: demos/tests/run_tests.py [hunk 5]
```diff
@@ -210,10 +212,26 @@ def main():
 
     if args.demos is not None:
         names_of_demos_to_test = set(args.demos.split(','))
+        if all(impl in Demo.IMPLEMENTATION_TYPES for impl in names_of_demos_to_test):
+            impl_types = names_of_demos_to_test
+            names_of_demos_to_test = set()
+            for impl in impl_types:
+                names_of_demos_to_test.update({demo.subdirectory for demo in DEMOS if demo.implementation == impl})
```
Comment (anzhella-pankratova) [inline demos/tests/run_tests.py:219]: the simplification:
```suggestion
            names_of_demos_to_test = {demo.subdirectory for demo in DEMOS if demo.implementation in names_of_demos_to_test}
```

**Comment by Wovchena** [comment]
@ivikhrev, now you need to add 3 jobs to precommits running cpp, gapi and python separately, keeping old one for now

**Pull request merged**